locations/spiders/safelite.py

```diff
@@ -0,0 +1,174 @@
+
+STATES = {
+    "AK": "Alaska",
+    "AL": "Alabama",
+    "AR": "Arkansas",
+    "AS": "American Samoa",
+    "AZ": "Arizona",
+    "CA": "California",
+    "CO": "Colorado",
+    "CT": "Connecticut",
+    "DC": "District of Columbia",
+    "DE": "Delaware",
+    "FL": "Florida",
+    "GA": "Georgia",
+    "GU": "Guam",
+    "HI": "Hawaii",
+    "IA": "Iowa",
+    "ID": "Idaho",
+    "IL": "Illinois",
+    "IN": "Indiana",
+    "KS": "Kansas",
+    "KY": "Kentucky",
+    "LA": "Louisiana",
+    "MA": "Massachusetts",
+    "MD": "Maryland",
+    "ME": "Maine",
+    "MI": "Michigan",
+    "MN": "Minnesota",
+    "MO": "Missouri",
+    "MS": "Mississippi",
+    "MT": "Montana",
+    "NC": "North Carolina",
+    "ND": "North Dakota",
+    "NE": "Nebraska",
+    "NH": "New Hampshire",
+    "NJ": "New Jersey",
+    "NM": "New Mexico",
+    "NV": "Nevada",
+    "NY": "New York",
+    "OH": "Ohio",
+    "OK": "Oklahoma",
+    "OR": "Oregon",
+    "PA": "Pennsylvania",
+    "PR": "Puerto Rico",
+    "RI": "Rhode Island",
+    "SC": "South Carolina",
+    "SD": "South Dakota",
+    "TN": "Tennessee",
+    "TX": "Texas",
+    "UT": "Utah",
+    "VA": "Virginia",
+    "VI": "Virgin Islands",
+    "VT": "Vermont",
+    "WA": "Washington",
+    "WI": "Wisconsin",
+    "WV": "West Virginia",
+    "WY": "Wyoming",
+}
+
+DAY_MAPPING = [
+    "Monday",
+    "Tuesday",
+    "Wednesday",
+    "Thursday",
+    "Friday",
+    "Saturday",
+    "Sunday",
+]
+
+
+class SafeliteSpider(scrapy.Spider):
+    name = "safelite"
+    allowed_domains = ["www.safelite.com"]
+    item_attributes = {"brand": "Safelite", "brand_wikidata": "Q28797369"}
+    start_urls = ("https://www.safelite.com/store-locator/store-locations-by-state",)
+
+    def parse(self, response):
+        state_urls = response.xpath(
+            './/div[@class="state-store-list"]/a/@href'
+        ).extract()
+        for state_url in state_urls:
+            yield scrapy.Request(
+                f"https://www.safelite.com{state_url}", callback=self.parse_location
+            )
+
+    def parse_location(self, response):
+        try:
+            addr_full = (
+                response.xpath('.//div[@class="content-block"]')[1]
+                .xpath(".//p/text()")
+                .extract_first()
+            )
+            addr_full = addr_full.split(",")
+            street_address = ", ".join(addr_full[:-2])
+            city = addr_full[-2].strip()
+            state_postcode = addr_full[-1].strip()
+            state, postcode = state_postcode.strip().split(" ")
+            state = state.strip()
+            state = STATES.get(state, state)
+            addr_full = ", ".join(addr_full)
+            lat = response.xpath(
+                './/div[@class="store-map"]/@data-start-lat'
+            ).extract_first()
+            lon = response.xpath(
+                './/div[@class="store-map"]/@data-start-lon'
+            ).extract_first()
+        except:
```

Comment:
If there's an error, we should let the exception bubble up instead of outputting some empty data.

Reply:
The exception is caused in cases where the store does not have any location data and only opening hours data. However, the HTML structure for this data is pretty much the same as the location data, without differences in any classes or tags. Hence, the only alternative here was to put it in a try catch block since pages without location selectors would fail the split clause and directly pick up the store timings.